### DISCUSSION TITLE
Cordinate maps stuff

### DISCUSSION THREAD

**Discussion by MrBago** [dipy/dipy#2361] [3. Community Lounge]
Hi all,
I've been writing some code for fiber tracking that I was hopping we could merge into the main branch soon, but before we do I wanted to document this coordinate map issue. I find coordinate map stuff very hard to explain clearly but I've tried hard to make it clear. To try and make things even more clear I've made the following script, https://github.com/MrBago/dipy/blob/master/scratch/coordmap_example.py Running this script and looking the the streamlines that result using trackvis will help clarify the discussion that follows. You probably need to check out my master branch of the dipy repository for the script to work.

My fiber tracking code uses what I will call the 'trackvis" coordinate mapping system. In this coordinate map one corner of the first voxel of the image is at [0, 0, 0]. The diagonal corner of the same voxel is at  [vs, vs, vs](assuming isotropic voxels) and the diagonal corner of the image is at [vs_x, vs_y, vs*z] where [x, y, z] is the shape of the image. Some existing code in dipy uses what I will call the "nifti" coordinate system, this is the coordinate system of a nifti image with an affine that has [vs, vs, vs, 1] along the diagonal and 0 elsewhere. I've drawn 2d examples of these coordinate systems bellow.

```
Trackvis:
A------------
| C |   |   |
----B--------
|   |   |   |
-------------
|   |   |   |
------------D

A = [0, 0]
B = [1, 1]
C = [.5, .5]
D = [3, 3]



Nifti:
A------------
| C |   |   |
----B--------
|   |   |   |
-------------
|   |   |   |
------------D

A = [-.5, -.5]
B = [.5, .5]
C = [0, 0]
D = [2.5, 2.5]
```

I'm bringing this up because I think it can be very confusing to have two coordinate systems in dipy. This is partially my fault because I wrongly assumed that dipy was already using the trackvis coordinate system when I started writing my code, but before I go changing a bunch of code I'd like to discuss the merits of these coordinate systems. The trackvis coordinate system seems to have at least two clear advantages over the nifti coordinate system. 

First, if we use the trackvis coordinate system in dipy, we can save our tracks directly to .trk files, without having to change coordinate systems first, and our tracks will be displayed correctly in trackvis.

Second, the trackvis coordinate system seems to be a 'natural' coordinate system for holding streamline information. That is to say this coordinate system is independent of the voxel size of the image volume. This is not true for the nifti coordinate system, and can be very useful for limiting the number of places where one needs to think about and deal with coordinate system changes.

**Comment by Garyfallidis** [comment]
Okay I went through the things that have advantages using the centre of the voxel as the origin and I realized that this is both very helpfull for registration and visualization of volumetric and tractographic data sets. Also it fits very well with how nibabel understands images. Therefore, it makes perfect sense to keep that coordinate system as the main dipy coordinate  system and then we will just need to add .5 every time we export in Trackvis format. Furthermore, the trk format is not the best format ever made it actually keeps more information of what is necessary and also it is very slow with big tractographies. So, +1 from me for keeping the origin on the center of the voxel! :-) Good catch though!!
